Source/Tests/LinuxSyscalls/Syscalls.cpp

```diff
@@ -321,26 +350,60 @@ uint64_t ExecveHandler(const char *pathname, char* const* argv, char* const* env
 
     auto OldArgv = argv;
 
-    // Skip filename argument
-    ++OldArgv;
-    while (*OldArgv) {
-      // Append the arguments together
-      ExecveArgs.emplace_back(*OldArgv);
+    // It is valid to provide nullptr first argument.
+    if (*OldArgv) {
+      // Skip filename argument
       ++OldArgv;
+      while (*OldArgv) {
+        // Append the arguments together
+        ExecveArgs.emplace_back(*OldArgv);
+        ++OldArgv;
+      }
+    }
+    else {
+      // Linux kernel will stick an empty argument in to the argv list if none are provided.
+       ExecveArgs.emplace_back(NullString);
     }
 
     // Emplace nullptr at the end to stop
     ExecveArgs.emplace_back(nullptr);
   }
 
-  if (Args) {
-    Result = ::syscall(SYS_execveat, Args->dirfd, "/proc/self/exe",
-                       const_cast<char *const *>(ExecveArgs.data()), envp, Args->flags);
-  }
-  else {
-    Result = execve("/proc/self/exe", const_cast<char *const *>(ExecveArgs.data()), envp);
+  if (IsFDExec) {
+    if (envp) {
+      auto OldEnvp = envp;
+      while (*OldEnvp) {
+        EnvpArgs.emplace_back(*OldEnvp);
+        ++OldEnvp;
+      }
+    }
+
+    int Flags = fcntl(Args.dirfd, F_GETFD);
+    if (Flags & FD_CLOEXEC) {
+      // We need to duplicate the FD if FD_CLOEXEC is set.
+      // FEX needs the FD to live past execve when binfmt_misc isn't used.
```

Comment:
```suggestion
      // FEX needs the FD to live past execve when binfmt_misc isn't used,
      // so duplicate the FD if FD_CLOEXEC is set.
```

An even better wording might be `Duplicate the FD if FD_CLOEXEC is set, so that FEX can...`

Reply:
Comment updated.